# 1.Jacob迭代法
### 代码：

In [1]:
import numpy as np
from numpy.linalg import inv,norm
import matplotlib.pyplot as plt
def jacobi(A,b,tol,K):
    D=np.diag(np.diag(A))
    U=(np.triu(A)-D)
    L=(np.tril(A)-D)
    err=1
    step=0
    x_old=np.zeros((len(b),1))
    x=np.zeros((len(b),1))
    while err>tol and step<K:
        x=inv(D).dot(b-(L+U).dot(x_old))
        err=norm(b-A.dot(x),2)/norm(b,2)
        x_old=x
        step+=1
    return x
N1=int(input("请输入Nx: "))
N2=int(input("请输入Ny: "))
K=int(input("请输入最大迭代步数K: "))
tol=float(input("请输入控制精度: "))
N=(N1+1)*(N2+1)
b=np.ones((N,1))
A=np.diag([4]*N)
for n in range(N):
    for m in range(N):
        if m==(n-N1-1) or m==(n+N1+1) or m==n-1 or m==n+1:
            A[n][m]=-1
print("解方程AX=b得X=")
print(jacobi(A,b,tol,K))

请输入Nx: 1
请输入Ny: 2
请输入最大迭代步数K: 100
请输入控制精度: 0.01
解方程AX=b得X=
[[0.89792095]
 [1.18075474]
 [1.41652536]
 [1.41652536]
 [1.18075474]
 [0.89792095]]


# 2.Gauss-Seidel迭代法
### 代码：

In [2]:
import numpy as np
from numpy.linalg import inv,norm
import matplotlib.pyplot as plt
def gauss_seidel(A,b,tol,K):
    D=np.diag(np.diag(A))
    U=(np.triu(A)-D)
    L=(np.tril(A)-D)
    err=1
    step=0
    x_old=np.zeros((len(b),1))
    x=np.zeros((len(b),1))
    while err>tol and step<K:
        x=np.dot(inv(inv(D).dot(L)+np.identity(len(b))),inv(D).dot(b-U.dot(x_old)))
        err=norm(b-A.dot(x),2)/norm(b,2)
        x_old=x
        step+=1
    return x
N=int(input("请输入N: "))
K=int(input("请输入最大迭代步数K: "))
tol=float(input("请输入控制精度: "))
b=np.ones((N,1))
for i in range(N):
    b[i][0]=i+1
A=np.zeros((N,N))
for i in range(N):
    for j in range(N):
        if j==i:
            A[i][j]=(i+1)**2
        else:
            A[i][j]=1/(i+j+3)
print("解方程AX=b得X=")
print(gauss_seidel(A,b,tol,K))

请输入N: 4
请输入最大迭代步数K: 100
请输入控制精度: 0.01
解方程AX=b得X=
[[0.79113818]
 [0.42959598]
 [0.30455659]
 [0.23554395]]


# 3.SOR迭代法
(1)取松弛因子w=1.1时，用SOR迭代求解题1


In [3]:
import numpy as np
from numpy.linalg import inv,norm
import matplotlib.pyplot as plt
def sor(A,b,tol,K,w):
    D=np.diag(np.diag(A))
    U=(np.triu(A)-D)
    L=(np.tril(A)-D)
    err=1
    step=0
    x_old=np.zeros((len(b),1))
    x=np.zeros((len(b),1))
    while err>tol and step<K:
        x=inv(w*L+D).dot((1-w)*D.dot(x_old)-w*U.dot(x_old))+w*inv(D+w*L).dot(b)
        err=norm(b-A.dot(x),2)/norm(b,2)
        x_old=x
        step+=1
    return x
N1=int(input("请输入Nx: "))
N2=int(input("请输入Ny: "))
K=int(input("请输入最大迭代步数K: "))
tol=float(input("请输入控制精度: "))
N=(N1+1)*(N2+1)
b=np.ones((N,1))
A=np.diag([4]*N)
for n in range(N):
    for m in range(N):
        if m==(n-N1-1) or m==(n+N1+1) or m==n-1 or m==n+1:
            A[n][m]=-1
print("解方程AX=b得X=")
print(sor(A,b,tol,K,1.1))

请输入Nx: 1
请输入Ny: 2
请输入最大迭代步数K: 100
请输入控制精度: 0.01
解方程AX=b得X=
[[0.89664351]
 [1.18100891]
 [1.41871105]
 [1.42039626]
 [1.18493124]
 [0.90156517]]


(2)当w取值为0，0.1，0.2，0.3……1.9，2时,求最优松弛因子

In [5]:
import numpy as np
from numpy.linalg import inv,norm
import matplotlib.pyplot as plt
def sor(A,b,tol,K,w):
    D=np.diag(np.diag(A))
    U=(np.triu(A)-D)
    L=(np.tril(A)-D)
    err=1
    step=0
    x_old=np.zeros((len(b),1))
    x=np.zeros((len(b),1))
    while err>tol and step<K:
        x=inv(w*L+D).dot((1-w)*D.dot(x_old)-w*U.dot(x_old))+w*inv(D+w*L).dot(b)
        err=norm(b-A.dot(x),2)/norm(b,2)
        x_old=x
        step+=1
    return step
N1=30
N2=30
K=2000
tol=1e-5
N=(N1+1)*(N2+1)
b=np.ones((N,1))
A=np.diag([4]*N)
for n in range(N):
    for m in range(N):
        if m==(n-N1-1) or m==(n+N1+1) or m==n-1 or m==n+1:
            A[n][m]=-1
step=list()
for w in np.linspace(0,2,21):
    step.append(sor(A,b,tol,K,w))
print(step)

[2000, 2000, 2000, 2000, 2000, 2000, 2000, 2000, 2000, 2000, 2000, 1930, 1572, 1269, 1010, 784, 587, 411, 253, 152, 2000]


当w取值为0，0.1，0.2，0.3……1.9，2时，迭代步数依次为

2000
2000
2000
2000
2000
2000
2000
2000
2000
2000
2000
1930
1572
1269
1010
784
587
411
253
152
2000

注意到，松弛因子取1.9时，迭代步数最少，为152，即w=1.9最优（环境为Thonny）